### ISSUE TITLE
Shipping time is required if we are syncing shipping rates

### ISSUE THREAD

**Issue by nima-karimi** [woocommerce/google-listings-and-ads#1380]
### Describe the bug:
Consider the following scenario; A user has stated that they will enter their shipping time settings in Google but they will enter their shipping rates manually or want them to be synced from WooCommerce shipping zones.

![image](https://user-images.githubusercontent.com/73110514/160776868-ce8c0cc6-8f22-45d6-a798-26c8bce8e7fe.png)

When we are submitting shipping rates to Merchant Center, the [deliveryTime](https://developers.google.com/shopping-content/reference/rest/v2.1/shippingsettings#deliverytime) is a required value. We expect this value to be set in the database when we sync the Merchant Center settings and if it's not in the database, syncing the shipping rates will fail.

### Steps to reproduce:

1. Choose the options mentioned above for the shipping time and shipping rate settings for your free listing
2. Save the settings so that they can be submitted to Google
3. Confirm that the sync fails due to a missing required value (deliveryTime) error.

### Expected behavior:
Shipping rates can't be synced without a deliveryTime; therefore, either the shipping time value must become required if the user has chosen the automatic or manual option for their shipping rates, or we need to fetch the shipping time from their Merchant Center settings if they have already set it up there.

### Actual behavior:
Syncing shipping rates fails if no estimated shipping time has been entered.

### Additional details:
A quick fix would be to make the shipping time input required on the UI if the user has chosen to manually enter their shipping rates or to sync them from their WooCommerce settings.

I don't think Google allows setting a global deliveryTime without assigning it to specific shipping services. This means that if the user chooses the "complex" option for the delivery time they also need to enter their shipping rates in Merchant Center. We might need to change the design on the shipping step so that these two are tied together.

![image](https://user-images.githubusercontent.com/73110514/160780567-45c50a5f-75ca-42aa-a445-e20d60425550.png)

cc: @j111q @ecgan

**Comment by ecgan** [comment]
> the shipping time value must become required if the user has chosen the automatic or manual option for their shipping rates

This is how I interpret it, or how I think the 🔘  radio button logic should be:

- When shipping rates are "automatic" or "simple", shipping times must be "simple". (We cannot have shipping times "complex", as stated in this issue.)
- When shipping rates are "complex", shipping times must be "complex". (I think it does not make sense to have shipping times "simple" and have them stored in GLA, when users have shipping rates "complex" and have that setup in Merchant Center. 🤔 )

> ... or we need to fetch the shipping time from their Merchant Center settings if they have already set it up there.

I think we shouldn't go that route. From my discussion with @j111q previously, ideally we want GLA to be the "source of truth" for most of the users. If we have the shipping rates in GLA and also "fetch the shipping time from their Merchant Center settings", there will be two sources of truth, I think it doesn't make sense from the user's point of view (two different places to maintain the two things), and I feel it is going to have some technical complications.

---

Given that this is a production issue, should we label this with `priority: high` and possibly come up with a quick fix? Though I don't think it is or would be a widespread common issue, as it is not very likely for the users to have that combination of radio options.

A quick fix that I can think of is to disable the shipping times options so that users can't change the options on their own. We can programmatically change the shipping times options based on the selected shipping rates options, using the 🔘  radio button logic I mentioned above. 

Tagging @j111q @jconroy for your input on this.

**Comment by findingsimple** [comment]
Getting late for me, I'll have to go digging in notes. I thought we previously had a default value or entry for shippingTime that we submitted. Maybe I'm getting this confused with another value

**Comment by findingsimple** [comment]
Does this block onboarding?

**Comment by nima-karimi** [comment]
> I thought we previously had a default value or entry for shippingTime that we submitted. Maybe I'm getting this confused with another value

No, we don't have any default values for this. We just leave it as null if no values are set in the database. 

> Does this block onboarding?

Yes, if we can't set a shipping rate then onboarding will fail with this error:

```

2022-03-30T12:14:34+00:00 ERROR Automattic\WooCommerce\GoogleListingsAndAds\API\Site\Controllers\MerchantCenter\{closure} {
  "error": {
    "code": 400,
    "message": "[services[0].deliveryTime] Missing required service delivery time",
    "errors": [
      {
        "message": "[services[0].deliveryTime] Missing required service delivery time",
        "domain": "global",
        "reason": "required"
      }
    ]
  }
}
```

**Comment by nima-karimi** [comment]
I've also noticed another issue: if you have previously set a value for the shipping times and now change the setting to "complex", the UI will still submit those values to the `POST mc/shipping/times/batch` API and the backend stores these values regardless of the option selected by the user.

**Comment by findingsimple** [comment]
> Yes, if we can't set a shipping rate then onboarding will fail with this error:

Ok well I'm  going to bump to critical

**Comment by findingsimple** [comment]
> No, we don't have any default values for this. We just leave it as null if no values are set in the database.

But previously, did we have shipping time value we used, I remember shopify did something like 99 days (we didn't do that) but I thought we did have a way of getting around it

**Comment by nima-karimi** [comment]
> But previously, did we have shipping time value we used, I remember shopify did something like 99 days (we didn't do that) but I thought we did have a way of getting around it

I think we decided not to set a default because that might confuse users and lead to unexpected results. Especially given that the estimated shipping time is displayed publicly right next to each listing product on Google. if I remember correctly, we also discussed this with Caio and he agreed it won't be a good idea.

**Comment by nima-karimi** [comment]
> I've also noticed another issue: if you have previously set a value for the shipping times and now change the setting to "complex", the UI will still submit those values to the POST mc/shipping/times/batch API and the backend stores these values regardless of the option selected by the user.

@ecgan Can you please check if this one has an easy fix too? I think this is an important issue and most likely the main reason why this bug has gone unnoticed for so long.

**Comment by j111q** [comment]
Thanks for flagging this. To fix it, I think @ecgan's suggestion makes sense:

> When shipping rates are "automatic" or "simple", shipping times must be "simple". (We cannot have shipping times "complex", as stated in this issue.)
When shipping rates are "complex", shipping times must be "complex".

I would recommend we don't _show_ the radio options for Shipping Rates though - since it's not something they can change or choose, I think it makes no sense to actually show the options.

I did a quick mock of what that would look like in [Figma](2c13f-pb): 

<img width="1457" alt="image" src="https://user-images.githubusercontent.com/65582817/160990670-68f2aee1-0587-44b3-a38b-a05e416bd57f.png">

<img width="1547" alt="image" src="https://user-images.githubusercontent.com/65582817/160990158-12ab763d-807f-4ab2-b8fe-f784d72e9d15.png">

I tweaked the copy of the third `Complex` option to say `My shipping settings are complex. I will enter my shipping rates and times manually in Google Merchant Center.`

Does that make sense?

**Comment by ecgan** [comment]
> > I've also noticed another issue: if you have previously set a value for the shipping times and now change the setting to "complex", the UI will still submit those values to the POST mc/shipping/times/batch API and the backend stores these values regardless of the option selected by the user.
> 
> @ecgan Can you please check if this one has an easy fix too? I think this is an important issue and most likely the main reason why this bug has gone unnoticed for so long.

@nima-karimi , I always thought that before those shipping rates and shipping times data are used, the backend API side would check on the `settings.shipping_rate` and `settings.shipping_time` first, as they are like a "master switch" / "parent switch", e.g. if `settings.shipping_time` is `simple`, only then we use the shipping time data (which is like a "child data"). Is it possible / not possible to do this on the API side? 🤔 

On the UI side, I think the intention was / is to retain the shipping times data, so that when users switch from `simple` > `complex` > `simple`, the shipping times data do not get deleted. For a "quick fix", I think I can find a way to save empty shipping times data to API when users select `complex`, while retaining the data on the front-end to continue to support options switching `simple` > `complex` > `simple`. However, personally I don't like that idea as a long term solution, as there is a "data mismatch" between UI and API. 

I'll look into implementing the quick fix above, together with @j111q's mock.

---

@j111q , thanks for the quick mock. I think it would take a bit more effort than my suggested quick fix in https://github.com/woocommerce/google-listings-and-ads/issues/1380#issuecomment-1082969598, as there are more UI changes and things moving around, but it is indeed a better UI and approach. I'll proceed based on your quick mock.

One question: should we combine the shipping rates and shipping times on the left hand side into one "Shipping rates and times" title? I think it make sense to combine them. 

![image](https://user-images.githubusercontent.com/417342/161003096-93d48234-307d-45d4-b4d5-9fc56d791d0a.png)

---

Since this issue is now labeled with `priority: critical`, I'll work on this issue as top priority. I think I might be able to get the code done earliest by tonight, latest by tomorrow Friday night.

**Comment by nima-karimi** [comment]
> e.g. if settings.shipping_time is simple, only then we use the shipping time data (which is like a "child data"). Is it possible / not possible to do this on the API side?

Currently, the API doesn't check anything but this should be easy to implement on the backend side. No need to change anything on the front-end then. I'll create a PR for this change soon.

**Comment by ecgan** [comment]
> I'll create a PR for this change soon.

Thank you @nima-karimi ! 🙌  I'm thinking the fixes (UI and API) for this issue should be released as a hotfix. I have created and pushed a branch `hotfix/1.12.1` that is based on the current trunk. Our PRs should then be based on and merged into the hotfix branch, and then we make a release from the hotfix branch when things are ready. 🤞

**Comment by nima-karimi** [comment]
I've created a PR for the backend changes here: https://github.com/woocommerce/google-listings-and-ads/pull/1386

I'll cherry-pick it to the hotfix branch as soon as it's merged to develop.

**Comment by nima-karimi** [comment]
https://github.com/woocommerce/google-listings-and-ads/pull/1386 is now merged to develop and cherry-picked into hotfix/1.12.1 branch

**Comment by ecgan** [comment]
This issue is fixed in PR #1386 and PR #1389. The fix is released in https://github.com/woocommerce/google-listings-and-ads/releases/tag/1.12.1. Closing this issue.
